thali/install/install.js

```diff
@@ -0,0 +1,240 @@
+}
+
+function createUnzipedDirectoryPath(depotName, branchName, directoryToInstallIn) {
+    return path.join(directoryToInstallIn, depotName + "-" + branchName);
+}
+
+function createGitHubZipResponse(depotName, branchName, directoryToInstallIn, directoryUpdated) {
+    return {
+        unzipedDirectory: createUnzipedDirectoryPath(depotName, branchName, directoryToInstallIn),
+        directoryUpdated: directoryUpdated
+    };
+}
+
+function installGitHubZip(projectName, depotName, branchName, directoryToInstallIn) {    
+    var gitHubZipUrlObject = getGitHubZipUrlObject(projectName, depotName, branchName);
+    
+    return doGitHubEtagsMatch(projectName, depotName, branchName, directoryToInstallIn)
+    .then(function(doTheEtagsMatch) {
+        if (doTheEtagsMatch) {
+            return createGitHubZipResponse(depotName, branchName, directoryToInstallIn, false);
+        }
+        return httpRequestPromise("GET", gitHubZipUrlObject)
+        .then(function(res) {
+            return new Promise(function(resolve, reject) {
+                res.pipe(unzip.Extract({ path: directoryToInstallIn}))
+                .on('close', function() {
+                    resolve();
+                }).on('error', function(e) {
+                   reject("Could not extract zip file " + gitHubZipUrlObject.href + ", error was " + e); 
+                });
+            }).then(function() {
+                return writeToEtagFile(depotName, branchName, directoryToInstallIn, res);
+            }).then(function() {
+                return createGitHubZipResponse(depotName, branchName, directoryToInstallIn, true);
+            });
+        });
+    });
+}
+
+function uninstallPluginsIfNecessary(weAddedPluginsFile, appRootDirectory) {
+    return fs.readFileAsync(weAddedPluginsFile).catch(function(err) {
+        if (err) {
+            if (err.code == fileNotFoundCode) {
+                return false;
+            }
+            
+            return Promise.reject(err);
+        }
+        
+        return true;
+    }).then(function(doWeNeedToUninstall) {
+        if (!doWeNeedToUninstall) {
+            return;
+        }
+        
+        return childProcessExecPromise('cordova plugin remove org.thaliproject.p2p', appRootDirectory)
+    })
+}
+
+/**
+ * This will copy the contents of a Thali_CordovaPlugin local depot to the right directory in the
+ * current Cordova project so it will be installed. This is used for local development only.
+ */
+function copyDevelopmentThaliCordovaPluginToProject(appRootDirectory, thaliDontCheckIn, depotName, branchName) {
+    var targetDirectory = createUnzipedDirectoryPath(depotName, branchName, thaliDontCheckIn);
+    var sourceDirectory = path.join(appRootDirectory, "../Thali_CordovaPlugin");
+    return new Promise(function(resolve, reject) {
+        fs.remove(targetDirectory, function(err) {
+            if (err) {
+                reject("copyDevelopmentThaliCordovaPluginToProject remove failed with " + err);
```

Comment:
The approach I took is that anytime I was sending the error string myself I wrapped it in "new error" but when I was passing on an error I got from someone else I passed it along as is.